quint/src/effects/base.ts

```diff
@@ -323,19 +324,21 @@ export function unifyEntities(va: Entity, vb: Entity): Either<ErrorTree, Substit
     return bindEntity(v1.name, v2).mapLeft(msg => buildErrorLeaf(location, msg))
   } else if (v2.kind === 'variable') {
     return bindEntity(v2.name, v1).mapLeft(msg => buildErrorLeaf(location, msg))
-  } else {
-    if (isEqual(v1, v2)) {
-      return right([])
-    }
-
-    if (v1.kind === 'union' && v2.kind === 'concrete') {
-      return mergeInMany(v1.entities.map(v => unifyEntities(v, v2)))
-        .map(subs => subs.flat())
-        .mapLeft(err => buildErrorTree(location, err))
-    }
-
-    if (v1.kind === 'concrete' && v2.kind === 'union') {
-      return unifyEntities(v2, v1)
+  } else if (isEqual(v1, v2)) {
+    return right([])
+  } else if (v1.kind === 'union' && v2.kind === 'concrete') {
+    return mergeInMany(v1.entities.map(v => unifyEntities(v, v2)))
+      .map(subs => subs.flat())
+      .mapLeft(err => buildErrorTree(location, err))
+  } else if (v1.kind === 'concrete' && v2.kind === 'union') {
+    return unifyEntities(v2, v1)
+  } else if (v1.kind === 'union' && v2.kind === 'union') {
+    if (intersectionWith(v1.entities, v2.entities, isEqual).length > 0) {
+      const s1 = { ...v1, entities: differenceWith(v1.entities, v2.entities, isEqual) }
+      const s2 = { ...v2, entities: differenceWith(v2.entities, v1.entities, isEqual) }
+
+      // There was an intersection, try to unify the remaining entities
+      return unifyEntities(s1, s2)
```

Comment:
Not sure if I'm following the code correctly, but won't this call just end up in the same branch (`v1.kind === 'union' && v2.kind === 'union'`) and then return the `left` below?

Reply:
Not if the `flattenUnions` calls in the first two lines of this function are able to perform some simplification. Might not be so obvious from the name, but `flattenUnions` actually transforms `[a]` in `a` (which are equivalent by definition, but the unification algorithm can only handle the latter).